File: app/src/main/java/space/serenity/berlinviewer/ui/providers/ReviewsProvider.java
package space.serenity.berlinviewer.ui.providers;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import space.serenity.berlinviewer.model.GYGReviewListResponse;
import space.serenity.berlinviewer.model.RestAPI;
import space.serenity.berlinviewer.model.Review;
import space.serenity.berlinviewer.ui.adapters.ReviewsAdapter;


/**
 * Created by karmamaker on 03.10.15.
 */
public class ReviewsProvider {
    private List<Review> source = new ArrayList<>();
    public static final int PAGE_SIZE = 32; // This value should be received from server
    public static final int ITEMS_BEFORE_NEXT_REQUEST = 24; // This value should be received from server
    protected int dataPadding; // Number of items before first item in source
    protected int lastLoadedPageSize = PAGE_SIZE;
    private ReviewsAdapter adapter;

    RestAPI api = new RestAPI(); // FIXME BS

    public ReviewsProvider(ReviewsAdapter adapter) {
        this.adapter = adapter;
    }

    protected Call<GYGReviewListResponse> currCall;

    public void init() {
        clear();
        currCall = requestPage(0, PAGE_SIZE, getCallback());
    }

    public void refresh() {
        init();
    }

    public void cancel(){
        if(currCall != null) {
            currCall.cancel();
            currCall = null;
        }
    }

    public void clear() {
        dataPadding = 0;
        lastLoadedPageSize = PAGE_SIZE;
    }

    public boolean isFullyLoaded(){
        return lastLoadedPageSize != PAGE_SIZE;
    }

    public Call<GYGReviewListResponse> requestPage(int pageNumber, int pageSize, Callback<GYGReviewListResponse> callback){
        Call<GYGReviewListResponse> call = api.getReviews(pageNumber, pageSize);
        call.enqueue(callback);
        return call;
    }

    protected Callback<GYGReviewListResponse> getCallback() {
        return new Callback<GYGReviewListResponse>() {
            @Override
            public void onResponse(Call<GYGReviewListResponse> call, Response<GYGReviewListResponse> response) {
                addPage(response.body().getData());
                notifyDataSetChanged();
                onFinish(call);
            }

            @Override
            public void onFailure(Call<GYGReviewListResponse> call, Throwable t) {
                onFinish(call);
            }

            private void onFinish(Call<GYGReviewListResponse> call) {
                currCall = null;
            }
        };
    }

    private void notifyDataSetChanged(){
        adapter.notifyDataSetChanged();
    }

    public Review get(int position) {
        int currentPage = (position + dataPadding) / PAGE_SIZE;

        if(shouldRequestNextPage(position)){
            currCall = requestPage(currentPage + 1, PAGE_SIZE, getCallback());
        }
        return source.get(position);
    }

    public int getCount() {
        return source.size();
    }

    private boolean shouldRequestNextPage(int position) {
        return !isFullyLoaded()
                && source.size() - ITEMS_BEFORE_NEXT_REQUEST < position
                && currCall == null;
    }

    protected void addPage(List loadedPage) {
        lastLoadedPageSize = loadedPage.size();
        int dataSizeWithDuplicates = loadedPage.size();
        loadedPage.removeAll(source);

        dataPadding += dataSizeWithDuplicates - loadedPage.size();
        source.addAll(source.size(), loadedPage);
    }
}

File: app/src/main/java/space/serenity/berlinviewer/ui/providers/ReviewsProvider.kt
package space.serenity.berlinviewer.ui.providers

import android.app.Notification
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import space.serenity.berlinviewer.model.GYGReviewListResponse
import space.serenity.berlinviewer.model.RestAPI
import space.serenity.berlinviewer.model.Review
import space.serenity.berlinviewer.ui.adapters.ReviewsAdapter
import java.util.ArrayList

/**
 * Created by karmamaker on 01/06/2017.
 */
class ReviewsProvider() {
    private val source = ArrayList<Any>()
    protected var dataPadding: Int = 0 // Number of items before first item in source
    protected var lastLoadedPageSize = PAGE_SIZE

    internal var api = RestAPI() // FIXME BS
    var dataSetChangeListener: () -> Unit = {}


    protected var currCall: Call<GYGReviewListResponse>? = null

    fun init() {
        clear()
        currCall = requestPage(0, PAGE_SIZE, callback)
    }

    fun clear() {
        dataPadding = 0
        lastLoadedPageSize = PAGE_SIZE
        val savedCall = currCall
        if(savedCall != null){
            savedCall.cancel()
            currCall = null
        }
    }

    val isFullyLoaded: Boolean
        get() = lastLoadedPageSize != PAGE_SIZE

    fun requestPage(pageNumber: Int, pageSize: Int, callback: Callback<GYGReviewListResponse>): Call<GYGReviewListResponse> {
        val call = api.getReviews(pageNumber, pageSize)
        call.enqueue(callback)
        return call
    }

    protected val callback: Callback<GYGReviewListResponse>
        get() = object : Callback<GYGReviewListResponse> {
            override fun onResponse(call: Call<GYGReviewListResponse>, response: Response<GYGReviewListResponse>) {
                if(response.body() != null) {
                    addPage(response.body().data)
                    notifyDataSetChanged()
                    onFinish(call)
                } else {
                    onFailure(call, Throwable("Bad response from server"))
                }

            }

            override fun onFailure(call: Call<GYGReviewListResponse>, t: Throwable) {
                if(isEmpty()){
                    source.clear()
                    source.add(ITEM_NO_CONNECTION_FULLSCREEN)
                } else {
                    source.add(ITEM_NO_CONNECTION_SMALL)
                }
                onFinish(call)
            }

            private fun onFinish(call: Call<GYGReviewListResponse>) {
                currCall = null
                source.remove(ITEM_LOADING)
            }
        }

    private fun notifyDataSetChanged() {
        dataSetChangeListener.invoke();
    }

    operator fun get(position: Int): Any {
        val currentPage = 0 //(position + dataPadding) / PAGE_SIZE FIXME

        if (shouldRequestNextPage(position)) {
            source.add(ITEM_LOADING)
            currCall = requestPage(currentPage + 1, PAGE_SIZE, callback)
        }
        return source[position]
    }

    val count: Int
        get() = source.size

    private fun shouldRequestNextPage(position: Int): Boolean {
        return !isFullyLoaded
                && source.size - ITEMS_BEFORE_NEXT_REQUEST < position
                && currCall == null
    }

    fun addPage(loadedPage: MutableList<Review>) {
        lastLoadedPageSize = loadedPage.size

        val dataSizeWithDuplicates = loadedPage.size
        loadedPage.removeAll(source)

        dataPadding += dataSizeWithDuplicates - loadedPage.size
        source.addAll(source.size, loadedPage)

    }

    fun isEmpty() : Boolean{
        return !source.any({it is Review })
    }

    companion object {
        val PAGE_SIZE = 32 // This value should be received from server
        val ITEMS_BEFORE_NEXT_REQUEST = 24 // This value should be received from server
        val ITEM_LOADING = "LOADING"
        val ITEM_NO_REVIEWS = "NO_REVIEWS"
        val ITEM_NO_CONNECTION_SMALL = "NO_CONNECTION_SMALL "
        val ITEM_NO_CONNECTION_FULLSCREEN = "NO_CONNECTION_FULLSCREEN"
    }
}